rpki-validator/src/main/java/net/ripe/rpki/validator3/domain/validation/ValidatedRpkiObjects.java

```diff
@@ -291,4 +258,70 @@ private void notifyListeners() {
         Locks.locked(dataLock.readLock(), () ->
             listeners.forEach(listener -> listener.accept(validatedObjectsByTrustAnchor.values())));
     }
+
+    public static class Accumulator {
+        private final TrustAnchorData trustAnchorData;
+        private final List<Key> validatedObjectKeys = Collections.synchronizedList(new ArrayList<>());
+        private final List<ValidatedRoaPrefix> validatedRoaPrefixes = Collections.synchronizedList(new ArrayList<>());
+        private final List<RouterCertificate> routerCertificates = Collections.synchronizedList(new ArrayList<>());
+
+        public Accumulator(TrustAnchorData trustAnchorData) {
+            this.trustAnchorData = trustAnchorData;
+        }
+
+        public void add(Key key, CertificateRepositoryObject object, ImmutableSortedSet<String> locations) {
+            validatedObjectKeys.add(key);
+            if (object instanceof RoaCms) {
+                RoaCms roa = (RoaCms) object;
+                for (RoaPrefix prefix: roa.getPrefixes()) {
+                    validatedRoaPrefixes.add(ValidatedRoaPrefix.of(
+                            trustAnchorData,
+                            roa.getAsn().longValue(),
+                            prefix.getPrefix(),
+                            prefix.getMaximumLength(),
+                            roa.getNotValidBefore().getMillis(),
+                            roa.getNotValidAfter().getMillis(),
+                            roa.getCertificate().getSerialNumber(),
+                            locations
+                    ));
+                }
+            } else if (object instanceof X509RouterCertificate) {
+                final Base64.Encoder encoder = Base64.getEncoder();
+                X509RouterCertificate certificate = (X509RouterCertificate) object;
+                final ImmutableList<String> asns = ImmutableList.copyOf(X509CertificateUtil.getAsns(certificate.getCertificate()));
+                final String ski = encoder.encodeToString(X509CertificateUtil.getSubjectKeyIdentifier(certificate.getCertificate()));
+                final String pkInfo = X509CertificateUtil.getEncodedSubjectPublicKeyInfo(certificate.getCertificate());
+                routerCertificates.add(RouterCertificate.of(
+                        trustAnchorData,
+                        asns,
+                        ski,
+                        pkInfo
+                ));
+            }
+        }
+
+        public boolean isEmpty() {
+            return validatedObjectKeys.isEmpty();
+        }
+
+        public int size() {
```

Comment:
Only works because it's O(1) on `ArrayList`, would be bad with my proposed change.